docs/layout-of-source-files.rst

```diff
@@ -216,101 +216,562 @@ the code below creates new global symbols ``alias`` and ``symbol2`` which refere
+   Only if the file actually exists under that path, it is added to the virtual filesystem and the
+   import key is based on the canonical path.
+
+#. **Standard JSON (as content)**
+
+   An alternative way to compile your project is to use the ``--standard-json`` option and provide
+   a JSON file containing all of your source code:
+
+   .. code-block:: json
+
+       {
+           "language": "Solidity",
+           "sources": {
+               "contract.sol": {
+                   "content": "import \"./util.sol\";\ncontract C {}"
+               },
+               "util.sol": {
+                   "content": "library Util {}"
+               },
+               "/usr/local/dapp-bin/token.sol": {
+                   "content": "contract Token {}"
+               }
+           },
+           "settings": {"outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}}
+       }
+
+   The ``sources`` dictionary specifies the initial content of the virtual filesystem and you
+   can use import keys directly there.
+   They do not undergo any extra translation or normalization.
+
+   The path to the JSON file does not affect the path resolution in any way.
+   In fact, it is common to supply it on the standard input in which case it does not have a path at all.
+
+   .. note::
+
+       When using ``--standard-json`` you cannot provide additional source files as command-line
+       arguments but it does not mean that the compiler will not load any extra files from disk.
+       If a contract imports a file that is not present in `sources`, the compiler will use the file
+       loader as in any other situation, which may result in the source being read from disk.
+
+#. **Standard JSON (as URL)**
+
+   When using Standard JSON it is possible to  tell the compiler to load the files from disk directly:
+
+   .. code-block:: json
+
+       {
+           "language": "Solidity",
+           "sources": {
+               "/usr/local/dapp-bin/token.sol": {
+                   "urls": ["/projects/mytoken.sol"]
+               }
+           },
+           "settings": {"outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}}
+       }
+
+   The path specified in ``urls`` is only passed to the file loader and used to locate the file.
+   It does not affect the import key and is not included in contract metadata.
+
+   Paths in ``urls`` are affected by base path and any other transformations performed by the file loader.
+
+#. **Standard input**
+
+   The last way to provide the source is by sending it to compiler's standard input:
+
+   .. code-block:: bash
+
+       echo 'import "./util.sol"; contract C {}' | solc -
+
+   The content of the standard input is identified in the virtual filesystem by a special import key:
+   ``<stdin>``.
+
+.. warning::
+
+    The compiler uses import keys to determine whether imports refer to the same module or not.
+    If you refer to a file in multiple ways that translate to different keys, it will be compiled
```

Comment:
This sounds like it would only generate the code twice, but the problem is that you might also get symbol collisions.

Reply:
I just checked and `--metadata` shows three different instances of the token contract.

You're right about the collision. I need to add the `as` part to the imports to avoid them.